zipkin-server/storage-cassandra/src/main/java/zipkin/server/storage/cassandra/dao/CassandraZipkinQueryDAO.java

```diff
@@ -0,0 +1,297 @@
+        .collect(toList());
+  }
+
+  @Override
+  public List<String> getRemoteServiceNames(String serviceName) throws IOException {
+    final Set<String> services = new HashSet<>();
+
+    for (String table : tableHelper.getTablesWithinTTL(ZipkinServiceRelationTraffic.INDEX_NAME)) {
+      services.addAll(client.executeQuery("select " + ZipkinServiceRelationTraffic.REMOTE_SERVICE_NAME +
+          " from " + table +
+          " where " + ZipkinServiceRelationTraffic.SERVICE_NAME + " = ?" +
+          " limit " + NAME_QUERY_MAX_SIZE,
+          row -> row.getString(ZipkinServiceRelationTraffic.REMOTE_SERVICE_NAME),
+          serviceName));
+    }
+
+    return services
+        .stream()
+        .limit(NAME_QUERY_MAX_SIZE)
+        .collect(toList());
+  }
+
+  @Override
+  public List<String> getSpanNames(String serviceName) throws IOException {
+    final Set<String> names = new HashSet<>();
+
+    for (String table : tableHelper.getTablesWithinTTL(ZipkinServiceSpanTraffic.INDEX_NAME)) {
+      names.addAll(client.executeQuery("select " + ZipkinServiceSpanTraffic.SPAN_NAME +
+          " from " + table +
+          " where " + ZipkinServiceSpanTraffic.SERVICE_NAME + " = ?" +
+          " limit " + NAME_QUERY_MAX_SIZE,
+          row -> row.getString(ZipkinServiceSpanTraffic.SPAN_NAME),
+          serviceName));
+    }
+
+    return names
+        .stream()
+        .limit(NAME_QUERY_MAX_SIZE)
+        .collect(toList());
+  }
+
+  @Override
+  public List<Span> getTrace(String traceId) {
+    final List<Span> spans = new ArrayList<>();
+
+    for (String table : tableHelper.getTablesWithinTTL(ZipkinSpanRecord.INDEX_NAME)) {
+      spans.addAll(client.executeQuery("select * from " + table +
+          " where " + ZipkinSpanRecord.TRACE_ID + " = ?" +
+          " limit " + NAME_QUERY_MAX_SIZE,
+          this::buildSpan, traceId));
+    }
+
+    return spans;
+  }
+
+  @Override
+  public List<List<Span>> getTraces(QueryRequest request, Duration duration) throws IOException {
+    Set<String> traceIdSet = new HashSet<>();
+    for (String table : tableHelper.getTablesForRead(
+        ZipkinSpanRecord.INDEX_NAME,
+        duration.getStartTimeBucket(),
+        duration.getEndTimeBucket()
+    )) {
+      List<CompletionStage<List<String>>> completionTraceIds = new ArrayList<>();
+      if (CollectionUtils.isNotEmpty(request.annotationQuery())) {
+        final long timeBucket = TableHelper.getTimeBucket(table);
+        final String tagTable = TableHelper.getTable(ZipkinSpanRecord.ADDITIONAL_QUERY_TABLE, timeBucket);
+        for (Map.Entry<String, String> entry : request.annotationQuery().entrySet()) {
+          completionTraceIds.add(client.executeAsyncQuery("select " + ZipkinSpanRecord.TRACE_ID + " from " + tagTable +
+                  " where " + ZipkinSpanRecord.QUERY + " = ?" +
+                  " and " + ZipkinSpanRecord.TIME_BUCKET + " >= ?" +
+                  " and " + ZipkinSpanRecord.TIME_BUCKET + " <= ? ALLOW FILTERING",
```

Comment:
is the primary key specified in the where clause here? 
allow filtering without the primary key specified is a full-table scan and will not work.

Reply:
I have updated to use of the latest secondary table in the v2 for searching.